### DISCUSSION TITLE
Confused with Tiptap Command Behaviour ?

### DISCUSSION THREAD

**Discussion by Guru4741** [ueberdosis/tiptap#2439] [General Questions]
I'm trying Tiptap editor and playing with some dummy code. But I'm confused with the behaviour. Could somebody please help me understand the behaviour ?

Below is the dummy code I'm trying (Below 4 cases were tried individually, just here for mentioning I'm placing them in one code block under same braces) :

```
const customCommand = () => ({commands, state, editor}) => {

// Doesn't work. Parent node doesn't change
console.log('current node ', state.selection.from)
commands.selectParentNode()
console.log('parent node ', state.selection.from)

// Doesn't work. Parent node doesn't change
console.log('current node ', state.selection.from)
this.editor.commands.selectParentNode()
console.log('parent node ', state.selection.from)

// Doesn't work. Parent node doesn't change
console.log('current node ', this.editor.view.state.selection.from)
commands.selectParentNode()
console.log('parent node ', this.editor.view.state.selection.from)

// Works. Parent Node changes.
console.log('current node ', this.editor.view.state.selection.from)
editor.commands.selectParentNode()
console.log('parent node ', this.editor.view.state.selection.from)

}
```

**Comment by philippkuehn** [comment]
Within a command you should use `tr.selection` instead of `state.selection` to get the current selection. 
And don’t forget to return a boolean at the end of your command to say if the command was successful.

```js
const customCommand = () => ({ commands, tr }) => {
  console.log('current node ', tr.selection.from)
  commands.selectParentNode()
  console.log('parent node ', tr.selection.from)

  return true
}
```

**Reply by Guru4741** [reply]
Thanks @philippkuehn for replying. I would use 'tr' in place of 'state' within commands. 

But I'm still confused why on using 'state' it works in some cases (the last one) and why it doesn't work in other cases ? Also could you please help me in knowing the difference between : this.editor.state.selection and state.selection within a command ? And are 'editor.commands.selectParentNode()' and 'commands.selectParentNode()' refers to two separate things ? I tried doing 'editor.commands.setBold()' and 'commands.setBold()' both resulted the same. But with selectParentNode() I faced issue.

I tried to go through the 'Tiptap' codebase, but didn't understand.

Thanks a lot.

**Reply by philippkuehn** [reply]
Everything is described on this page: https://tiptap.dev/api/commands

To support chainable commands, you must not use `editor.commands` within a command, since this would dispatch a transaction. The idea of supporting chainable commands is that you can run multiple commands while dispatching one single transaction (one update event).

But of course you can use a simple function instead of writing a custom command. This would result in three update events.

```js
function someMethod() {
  editor.commands.insertText('foo')
  editor.commands.insertText('foo')
  editor.commands.insertText('foo')
}
```
